### PR TITLE
TW-118: liquidity baking add apy and tr history

### PR DESCRIPTION
**Pull request by letier3110**

### CODE REVIEW

File: src/components/activity-groups-list/activity-group-item/activity-group-item.tsx [hunk 1]
```diff
@@ -21,7 +22,9 @@ interface Props {
 
 export const ActivityGroupItem: FC<Props> = ({ group }) => {
   const styles = useActivityGroupItemStyles();
-  const nonZeroAmounts = useNonZeroAmounts(group);
+  const hashGroup = useHashOperation(group[0].hash);
+
+  const nonZeroAmounts = useNonZeroAmounts(hashGroup[0] ?? group);
```
Comment (unixvb) [inline src/components/activity-groups-list/activity-group-item/activity-group-item.tsx:27]: make this component pure - let it receive already prepared data

File: src/hooks/use-hash-operation.hook.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    loadHashOperation();
+  }, [loadHashOperation]);
+
+  return activities;
+};
+
+const loadContractActivity = async (publicKeyHash: string, hash: string): Promise<Array<ActivityInterface>> => {
+  const operations = await getOperationGroupByHash<OperationLiquidityBakingInterface>(hash).then(x => x.data);
+
+  return mapOperationsToActivities(publicKeyHash, operations);
```
Comment (unixvb) [inline src/hooks/use-hash-operation.hook.ts:32]: you could just continue `.then()` chaining

**Review by LendiHop — APPROVED** [review]

File: src/hooks/use-contract-activity.ts [hunk 3]
```diff
@@ -49,12 +49,7 @@ const loadContractActivity = async (
...
-  const operations = await getContractOperations<OperationLiquidityBakingInterface>(
-    publicKeyHash,
-    contractAddress,
-    lastLevel
-  ).then(x => x.data);
-
-  return mapOperationLiquidityBakingToActivity(publicKeyHash, operations);
-};
+): Promise<Array<ActivityInterface>> =>
+  await getContractOperations<OperationLiquidityBakingInterface>(publicKeyHash, contractAddress, lastLevel).then(x =>
```
Comment (unixvb) [inline src/hooks/use-contract-activity.ts:53]: you could remove `await`

File: src/utils/array.utils.ts [hunk 4]
```diff
@@ -8,3 +8,6 @@ export const sliceIntoChunks = <T>(array: T[], chunkSize: number) => {
 
   return result;
 };
+
+export const deduplicate = <T>(arr: T[], filterPredicate: (a: T, b: T) => boolean) =>
+  arr.filter((value, index, self) => index === self.findIndex(t => filterPredicate(t, value)));
```
Comment (unixvb) [inline src/utils/array.utils.ts:13]: just use `import { uniq } from 'lodash-es';`

File: src/utils/token-operations.util.ts [hunk 5]
```diff
@@ -4,6 +4,8 @@ import { ActivityTypeEnum } from '../enums/activity-type.enum';
 import { OperationFa12Interface, OperationFa2Interface, OperationInterface } from '../interfaces/operation.interface';
 import { isDefined } from './is-defined';
 
+export const getOperationGroupByHash = <T>(hash: string) => tzktApi.get<Array<T>>(`operations/${hash}`);
```
Comment (unixvb) [inline src/utils/token-operations.util.ts:7]: it's extremely hard to read all those queries below.  
rewritethem using axios `config.params`

File: src/hooks/use-tezos-token-activity.hook.ts [hunk 6]
```diff
@@ -18,9 +20,19 @@ export const useTezosTokenActivity = (): UseActivityInterface => {
...
+      const operationGroups = (
+        await Promise.all(
+          filteredOperations.map(x =>
+            getOperationGroupByHash<OperationLiquidityBakingInterface>(x.hash).then(x => x.data)
+          )
+        )
+      ).flat();
+
+      const loadedActivities = mapOperationsToActivities(publicKeyHash, operationGroups);
       const activityGroups = transformActivityInterfaceToActivityGroups(loadedActivities);
```
Comment (unixvb) [inline src/hooks/use-tezos-token-activity.hook.ts:36]: it makes no sense that `operationGroups` data is `flatted` and then is grouped by `hash`.  

look like you don't need `transformActivityInterfaceToActivityGroups` at all (you could make activity groups at the moment of fetching  `operationGroups`).

File: src/hooks/use-contract-activity.ts [hunk 7]
```diff
@@ -1,40 +1,62 @@
...
+        publicKeyHash,
+        contractAddress,
+        tokenId,
+        tokenType,
+        setActivities,
+        setIsAllLoaded
+      });
+    }
+
+    return emptyFn;
```
Comment (unixvb) [inline src/hooks/use-contract-activity.ts:35]: you could just delete this line of code

File: src/utils/token-operations.util.ts [hunk 8]
```diff
@@ -1,85 +1,241 @@
...
+  if (operations.length === 0) {
+    return [];
+  }
+  const localLastItem = operations[operations.length - 1];
+  if (!isDefined(localLastItem)) {
+    return [];
+  }
+  const lowerId = localLastItem.id;
+  const operationsFa12 = await getFa12IncomingOperations(selectedRpc, publicKeyHash, lowerId, lastLevel);
+  const operationsFa2 = await getFa2IncomingOperations(selectedRpc, publicKeyHash, lowerId, lastLevel);
```
Comment (unixvb) [inline src/utils/token-operations.util.ts:163]: these could be run in parallel

File: src/hooks/use-contract-activity.ts [hunk 9]
```diff
@@ -1,42 +1,78 @@
...
+          return LoadLastActivityTokenType.Tezos;
+        case LIQUIDITY_BAKING_DEX_ADDRESS:
+          return LoadLastActivityTokenType.LiquidityBaking;
+        default:
+          return tokenType;
+      }
+    } else {
+      return LoadLastActivityTokenType.All;
+    }
+  }, [tokenType]);
```
Comment (unixvb) [inline src/hooks/use-contract-activity.ts:36]: delete `LoadLastActivityTokenType`, you don't need it.  
make this switch case where you are loading operations hashes

File: src/hooks/use-contract-activity.ts [hunk 10]
```diff
@@ -1,42 +1,78 @@
...
+    if (!loading) {
+      loadLastActivity({
+        selectedRpc,
+        lastLevel: null,
+        publicKeyHash,
+        contractAddress,
+        tokenId: tokenId ?? '0',
+        activityType,
+        setActivities,
+        setIsAllLoaded
```
Comment (unixvb) [inline src/hooks/use-contract-activity.ts:48]: do not send setters inside `loadLastActivity` let is return list of activities and do whatever you want with it

File: src/utils/token-operations.util.ts [hunk 11]
```diff
@@ -1,85 +1,243 @@
...
+        entrypoint: 'transfer',
+        'parameter.to': account,
+        ...(isDefined(upperId) ? { 'id.lt': upperId } : undefined)
+      }
+    })
+    .then(x => x.data);
 
 export const getFa2IncomingOperations = async (
-  selectedRpcUrl: string,
+  selectedRpc: string,
```
Comment (unixvb) [inline src/utils/token-operations.util.ts:129]: rename back, as it was (`selectedRpcUrl`).  
related to all fn above

File: src/utils/token-operations.util.ts [hunk 12]
```diff
@@ -1,85 +1,243 @@
...
+  const operationGroups = await Promise.all(
+    filteredOperations.map(x => getOperationGroupByHash<OperationInterface>(selectedRpc, x.hash).then(x => x.data))
+  );
+
+  const activityGroups: Array<ActivityGroup> = [];
+  for (const group of operationGroups) {
+    activityGroups.push(mapOperationsToActivities(publicKeyHash, group));
+  }
+
+  setActivities((prevValue: Array<ActivityGroup>) => [...prevValue, ...activityGroups]);
```
Comment (unixvb) [inline src/utils/token-operations.util.ts:242]: just return mapped  `operationGroups` array

File: src/utils/token-operations.util.ts [hunk 13]
```diff
@@ -1,85 +1,243 @@
...
+      throw new Error('unimplemented');
+  }
+
+  const filteredOperations = uniqBy<OperationInterface>(operations, 'hash');
+
+  setIsAllLoaded(filteredOperations.length === 0);
+
+  const operationGroups = await Promise.all(
+    filteredOperations.map(x => getOperationGroupByHash<OperationInterface>(selectedRpc, x.hash).then(x => x.data))
+  );
```
Comment (unixvb) [inline src/utils/token-operations.util.ts:235]: you could map data over here

**Review by oryhivskiy — APPROVED** [review]

**Review by NovakVlad2424 — APPROVED** [review]

**Pull request merged**